reasoner/controller/ConjunctionController.java

```diff
@@ -184,80 +194,115 @@ NegatedRequest asNegated() {
-            private final ConceptMap initialPacket;
+            private final ConjunctionStreamPlan plan;
+            private final ConceptMap identifierBounds;
             private final AbstractProcessor<?, ?, ?, ?> processor;
+            private final Map<Publisher<ConceptMap>, Integer> whichChild;
 
-            CompoundStream(AbstractProcessor<?, ?, ?, ?> processor, List<Resolvable<?>> plan,
-                           ConceptMap initialPacket) {
-                super(processor, new SubscriberRegistry.Single<>(), new PublisherRegistry.Multi<>());
+
+            CompoundStream(AbstractProcessor<?, ?, ?, ?> processor, ConjunctionStreamPlan plan, ConceptMap identifierBounds) {
+                super(processor, new SubscriberRegistry.Multi<>(), new PublisherRegistry.Multi<>());
                 this.processor = processor;
-                assert plan.size() > 0;
-                this.initialPacket = initialPacket;
-                this.remainingPlan = new ArrayList<>(plan);
-                this.publisherPackets = new HashMap<>();
-                this.leadingPublisher = nextCompoundLeader(this.remainingPlan.remove(0), initialPacket);
-                this.leadingPublisher.registerSubscriber(this);
+                this.identifierBounds = identifierBounds;
+                this.plan = plan;
+                Publisher<ConceptMap> leftChild;
+                if (plan.isCompoundStream()) {
+                    leftChild = spawnPlanElement(plan.asCompoundStreamPlan().ithChild(0), identifierBounds);
+                } else {
+                    leftChild = spawnPlanElement(plan.asResolvablePlan(), identifierBounds);
+                }
+                leftChild.registerSubscriber(this);
+                this.whichChild = new HashMap<>();
+                this.whichChild.put(leftChild, 0);
                 processor().context().perfCounters().compoundStreams.add(1);
             }
 
+            private Publisher<ConceptMap> spawnPlanElement(ConjunctionStreamPlan planElement, ConceptMap bounds) {
+                ConceptMap extension = filterOutputsWithExplainables(bounds, planElement.extendOutputWithVariables());
+                ConceptMap identifiers = bounds.filter(planElement.identifierVariables());
+                assert planElement.identifierVariables().size() == identifiers.concepts().size() &&  identifiers.concepts().keySet().containsAll(planElement.identifierVariables());
+                Publisher<ConceptMap> publisher;
+                if (planElement.isResolvable()) {
+                    publisher = spawnResolvableElement(planElement.asResolvablePlan(), identifiers);
+                } else if (planElement.isCompoundStream()) {
+                    publisher = spawnCompoundStream(planElement.asCompoundStreamPlan(), identifiers);
+                } else throw TypeDBException.of(ILLEGAL_STATE);
+
+                return extendWithBounds(publisher, extension);
```

Comment:
If a variable that's been found by a previous resolvable/subtree was not relevant to this subtree, we don't pass it down to our subtree but keep it in the compound stream and extend any answer from our subtree with it.

(These are the diamond join nodes in the figures we drew)